File: src/checkers-python/StudentAI.py
from random import randint
from tracemalloc import start
from BoardClasses import Move
from BoardClasses import Board
#The following part should be completed by students.
#Students can modify anything except the class name and exisiting functions and varibles.
import math
# import time as t

# What is a leaf node? : https://ai.stackexchange.com/questions/25426/unclear-definition-of-a-leaf-and-diverging-utc-values-in-the-monte-carlo-tree
# Some struggles - what is a leaf node? How to define it in the context of MCTS
#                - Best parameters for C and ITERATIONS
#                - Should I implement a heuristic?
#                - How to improve runtime efficiency?

C = math.sqrt(2)
ITERATIONS = 500

class Node():
    def __init__(self, board_history, color, parent, unexplored_children, move=None):
        self.board_history = board_history
        self.move = move
        self.color = color
        self.parent = parent
        self.unexplored_children = unexplored_children
        self.unexplored_children_length = len(unexplored_children)

        self.s_i = 0
        self.w_i = 0 
        self.children = []

    def update_history(self, move):
        self.board_history.append(move)



class StudentAI():
    def __init__(self,col,row,p):
        self.col = col
        self.row = row
        self.p = p
        self.board = Board(col,row,p)
        self.board.initialize_game()
        self.color = ''
        self.opponent = {1:2,2:1}
        self.color = 2
        
        # My code
        self.root = None


    def uct_score(self, node):
        return (node.w_i/node.s_i) + C*(math.sqrt(math.log(node.parent.s_i)/node.s_i))

    def get_move(self,move):
        if len(move) != 0:
            self.board.make_move(move,self.opponent[self.color])
        else:
            self.color = 1

        self.board.saved_move = []
        get_all_possible_moves = [move for move_list in self.board.get_all_possible_moves(self.color) for move in move_list]
        self.root = Node([], self.color, None, get_all_possible_moves, None)
            
        i = 0
        while (i < ITERATIONS):
            to_expand = self.selection(self.root)
            child = self.expand(to_expand)
            winner_color = self.simulation(child)
            self.backpropogate(child, winner_color)            
            i += 1
        
        curr_max_val = float('-inf') 
        best_move = None
        for node in self.root.children:
            if node.s_i > curr_max_val:
                curr_max_val = node.s_i
                best_move = node.move
        self.board.make_move(best_move, self.color)

        return best_move

    def update_board(self, root):
        color = self.color # Since starting from self.root.
        for move in root.board_history: # Update board w/ root_history
            self.board.make_move(move, color)
            color = self.opponent[color]
    
    def undo_board(self):
        while self.board.saved_move != []:
            self.board.undo()
            
    def selection(self, root):
        if root.children == [] or root.unexplored_children_length != 0:
            return root
        else:
            best_score = float('-inf')
            for child in root.children:
                score = self.uct_score(child)
                if score > best_score:
                    best_score = score
                    node = child
            return self.selection(node)

    def expand(self, root):
        if root.unexplored_children_length == 0:
            # This is a terminal node.
            return root

        self.update_board(root)

        index = randint(0, root.unexplored_children_length-1)
        move = root.unexplored_children.pop(index)
        root.unexplored_children_length -= 1

        new_history = list(root.board_history)
        new_history.append(move)

        self.board.make_move(move, root.color)
        new_color = self.opponent[root.color]

        get_all_possible_moves = [move for move_list in self.board.get_all_possible_moves(new_color) for move in move_list]
        node = Node(new_history, new_color, root, get_all_possible_moves, move)

        root.children.append(node)

        return node
        
    def simulation(self, root):
        board = self.board

        turn_table = {1: "W", 2: "B"}
        player = root.color

        while True:

            winner = board.is_win(turn_table[player])
            if winner != 0:
                break

            moves = board.get_all_possible_moves(player)
            winner = board.is_win(turn_table[player])
            print(winner, player, moves)
            index = randint(0, len(moves)-1)
            inner_index = randint(0, len(moves[index])-1)
            move = moves[index][inner_index]
            board.make_move(move, player)


            player = self.opponent[player]

        self.undo_board()
        return winner
    
    def backpropogate(self, root, winner):
        while root != None:
            root.s_i += 1 # Always increment
            if root.color != winner:
                root.w_i += 1
            
            root = root.parent

File: src/checkers-cpp/StudentAI.cpp
#include "StudentAI.h"
#include <random>
#include <limits>

//The following part should be completed by students.
//The students can modify anything except the class name and exisiting functions and varibles.
double C = sqrt(2);
int ITER = 500;

Node::Node(vector<Move> boardHist, int color, Node* parent, vector<Move> unexploredChildren, Move* move)
    :boardHist{boardHist}, color{color}, parent{parent}, unexploredChildren{unexploredChildren}, move{move}
{
    children;
    unexploredChildrenLen = unexploredChildren.size();

    s_i = 0;
    w_i = 0;
}

Node::~Node()
{
    for (int i = 0; i < this->children.size(); ++i)
    {
        delete this->children[i]->move;
        delete this->children[i];
    }
}

void Node::deallocate(Node* node)
{
    // for (int i = 0; i < node->children.size(); ++i)
    // {
    //     delete node->children[i]->move;
    //     delete node->children[i];
    // }
}

void Node::printChildren()
{
    cout << "  Children" << endl;
    for (int i = 0; i < children.size(); ++i)
    {
        cout << "    " << children[i] << endl;
    }
}

void Node::addHistory(Move move)
{
    boardHist.push_back(move);
}

StudentAI::StudentAI(int col,int row,int p)
	:AI(col, row, p)
{
    board = Board(col,row,p);
    board.initializeGame();
    player = 2;
    root = NULL;
}

vector<Move> flattenVect(vector<vector<Move>> moves)
{
    vector<Move> flattenedVec;
    for (int i = 0; i < moves.size(); ++i)
    {
        for (int j = 0; j < moves[i].size(); ++j)
        {
            flattenedVec.push_back(moves[i][j]);
        }
    }
    return flattenedVec;
}

Move StudentAI::GetMove(Move move)
{
    if (move.seq.empty())
    {
        player = 1;
    } else{
        board.makeMove(move,player == 1?2:1);
    }
    
    board.saved_move_list.clear();
    vector<vector<Move>> getAllMoves = board.getAllPossibleMoves(player);

    vector<Move> emptyVector;

    Node* parent;
    root = new Node(emptyVector, player, NULL, flattenVect(getAllMoves), NULL); 
    
    int i = 0;
    Node* toExpand;
    Node* child;
    int winner;

    while (i < ITER)
    {
        // cout << "Loop: " << i << endl;
        toExpand = selection(root);
        // cout << "1" << endl;
        child = expand(toExpand);
        // cout << "2" << endl;
        winner = simulation(child);
        // cout << "3" << endl;
        backpropagate(child, winner);
        // cout << "4" << endl;
        ++i;
    }

    // cout << "loop finished" << endl;
    double currMax  = -INFINITY;
    Node* bestMove = NULL;
    root->children.size();
    for (int i = 0; i < root->children.size(); ++i)
    {
        // cout << i << endl;
        if ((root->children[i])->s_i > currMax)
        {
            currMax = root->s_i;
            bestMove = root->children[i];
        }
    }
    // cout << "loop finished" << endl;
    // cout << bestMove->move->toString() << endl;
    Move bestMv = *(bestMove->move);
    board.makeMove(bestMv, player);
    // board.showBoard();
    delete root;
    return bestMv;
}

void StudentAI::updateBoard(Node* node)
{
    int p = player;
    // board.showBoard();
    for (int i = 0; i < node->boardHist.size(); ++i)
    {
        // cout << p << " & " << node->boardHist[i].toString() << endl;
        board.makeMove(node->boardHist[i], p);
        p = (p == 1?2:1);
    }
}
void StudentAI::undoBoard()
{
    while (board.saved_move_list.size() != 0)
    {
        board.Undo();
    }
}
double StudentAI::uctScore(Node* node)
{
    // return (node.w_i/node.s_i) + C*(math.sqrt(math.log(node.parment.s_i)/node.s_i))
    return (node->w_i / node->s_i) + C*(sqrt(log(node->parent->s_i/node->s_i)));
}

Node* StudentAI::selection(Node* node)
{
    if (node->children.size() == 0 || node->unexploredChildrenLen != 0)
    {
        return node;
    }
    else
    {
        Node* best_node = NULL;
        double best_score = -INFINITY;
        double score;
        for (int i = 0; i < node->children.size(); ++i)
        {
            score = uctScore(node->children[i]);
            // cout << i << ": " << node->children[i] << " " << node->s_i << " " <<  node->w_i << endl;
            
            if (score > best_score)
            {
                best_score = score;
                best_node = node->children[i];
            }
        }
        // cout << "Best: " << best_node << endl;
        // cout << best_node->children.size() << endl;
        // cout << best_node->unexploredChildrenLen << endl;
        return selection(best_node);

    }
}

Node* StudentAI::expand(Node* node)
{
    if (node->unexploredChildrenLen == 0)
    {
        return node;
    }

    updateBoard(node);

    int index = rand() % (node->unexploredChildrenLen);
    Move move = (node->unexploredChildren)[index];
    node->unexploredChildren.erase(node->unexploredChildren.begin()+index);
    node->unexploredChildrenLen--;

    vector<Move> newHistory; // Check if this copy is correct
    for (int i = 0; i < node->boardHist.size(); ++i)
    {
        newHistory.push_back(node->boardHist[i]);
    }
    newHistory.push_back(move);

    board.makeMove(move, node->color);
    int newPlayer = (node->color==1?2:1);
    Move* saveMove = new Move(move);

    Node* newNode = new Node(newHistory, newPlayer, node, flattenVect(board.getAllPossibleMoves(newPlayer)), saveMove);

    node->children.push_back(newNode);
    return newNode;
}

int StudentAI::simulation(Node* node)
{
    int player = root->color;
    int winner;
    string p;
    vector<vector<Move>> moves;
    int outer_index;
    int inner_index;
    Move move;

    while (true)
    {
        if (player == 1)
        {
            p = "W";
        }
        else
        {
            p = "B";
        }
        winner = board.isWin(p);

        if (winner != 0)
        {
            break;
        }

        moves = board.getAllPossibleMoves(player);
        outer_index = rand() % moves.size();
        inner_index = rand() % moves[outer_index].size();
        move = moves[outer_index][inner_index];

        board.makeMove(move, player);
        player = (player == 1?2:1);

    }
    undoBoard();
    return winner;

}

void StudentAI::backpropagate(Node* node, int winner)
{
    while (node != NULL)
    {
        node->s_i++;
        if (node->color != winner)
        {
            root->w_i++;
        }
        node = node->parent;
    }
}

